File: leetcode/clone_graph.js
/* Clone an undirected graph. 
 * Each node in the graph contains a label and a list of its neighbors.
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

/**
 * Definition for undirected graph.
 * function UndirectedGraphNode(label) {
 *     this.label = label;
 *     this.neighbors = [];   // Array of UndirectedGraphNode
 * }
 */

/**
 * @param {UndirectedGraphNode} graph
 * @return {UndirectedGraphNode}
 */
var cloneGraph = function(graph) {
    function helper(node, map) {
        var new_node = new UndirectedGraphNode(node.label);
        map.set(node.label, new_node);
        for (var neighbor of node.neighbors) {
            if (!map.get(neighbor.label))
                map.set(neighbor.label, helper(neighbor, map));
            new_node.neighbors.push(map.get(neighbor.label));
        }
        return new_node;
    }
    return graph ? helper(graph, new Map()) : null;
};

File: lintcode/clone_graph.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []
class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        def helper(node, map):
            new = UndirectedGraphNode(node.label)
            map[node.label] = new
            for neighbor in node.neighbors:
                if neighbor.label not in map:
                    map[neighbor.label] = helper(neighbor, map)
                new.neighbors.append(map[neighbor.label])
            return new
        return helper(node, {}) if node else None
